### ISSUE TITLE
Investigate places that EvaluationContext should be used

### ISSUE THREAD

**Issue by davkean** [dotnet/project-system#4649]
CPS introduced a concept called IProjectEvaluationContextService that enables features throughout CPS to turn usage of a "shared" evaluation context in MSBuild so that disk lookups (for globs) and SDK resolves are shared between different project evaluations. This feature was never turned on (and hidden behind remote setting) because no benefit was found at the time. I believe that https://github.com/Microsoft/msbuild/issues/4025 might change this.

We should do a few things:

- [ ] Investigate whether we should turn this on for solution load
- [ ] Investigate whether we should turn this on places that we bulk re-evaluate - such as in CPS's ProjectReevaluatingTriggerService

**Comment by japj** [comment]
I encountered an issue with VisualStudio 2017 15.8.1 where Sdk resolution seems to fail for (initial) unclear reasons. Reading up on this, I think this explains the behaviour.

We have a NuGet.Config and global.json in the root of the source tree to specify a TFS PackageManagement feed with a custom MSBuild Sdk.

This MSBuild Sdk contains some targets that generate project/code files out of source tree.
(the generated projects also reference this MSBuild Sdk). The source tree contains a project that causes these targets to run and uses the MSBuild task to build the generated project.

The Sdk resolution fails from Visual Studio for this generated project (it works ok when running MSBuild from command line).

If I copy the NuGet.config and global.json to the generated folder it works ok from VisualStudio.

I guess this behaviour is related to not having a shared MSBuild context across projects from VisualStudio?

**Comment by davkean** [comment]
Evaluation context isn't turned on currently in VS, so it won't/shouldn't be explained by that. 

I'm having troubling picturing the setup though, do you have a repro?

**Comment by japj** [comment]
Ok, I finally have been able to create an isolated case to show the behaviour I tried describing at https://github.com/japj/ShowSdkResolverBehaviour

My impression now is that this might be caused by how the c++ project build system and the dotnet project build system inside Visual Studio behave differently.

Perhaps (not) sharing MSBuild context information or (totally guessing here) the c++ build inside Visual Studio was not updated to include the SdkResolver mechanism as part of supporting MSBuild 15 in VisualStudio 2017, while the new project-system did?

I hope this helps pinpoint this actual problem? (and perhaps move it to the 'right' owner inside Microsoft?)

**Comment by japj** [comment]
@davkean @rainersigwald should I submit the issue related to the VS SDK Resolving differences from C++ projects somewhere separately? (I don’t want it to be lost as a comment here above)

**Comment by rainersigwald** [comment]
Yes, please file a feedback issue. That will file a bug we can use internally to figure out the root cause. Off the top of my head I have no idea why that would be different in vcxproj.

**Comment by japj** [comment]
@rainersigwald done at https://developercommunity.visualstudio.com/content/problem/537164/visualstudio-msbuild-fails-to-resolve-sdk-with-vcx.html
